eclair-core/src/main/scala/fr/acinq/eclair/message/OnionMessages.scala

```diff
@@ -0,0 +1,108 @@
+package fr.acinq.eclair.message
+
+import fr.acinq.bitcoin.Crypto.{PrivateKey, PublicKey}
+import fr.acinq.eclair.crypto.Sphinx
+import fr.acinq.eclair.wire.protocol.MessageOnion.{finalBlindedTlvCodec, messageOnionPerHopPayloadCodec, messageRelayPayloadCodec, relayBlindedTlvCodec}
+import fr.acinq.eclair.wire.protocol.MessageTlv._
+import fr.acinq.eclair.wire.protocol.{BadOnion, EncryptedRecipientDataTlv, OnionMessage, OnionMessagePayloadTlv, TlvStream}
+import scodec.bits.ByteVector
+import scodec.{Attempt, DecodeResult}
+
+import scala.util.{Failure, Success}
+
+object OnionMessages {
+
+  case class IntermediateNode(nodeId: PublicKey, padding: Option[ByteVector] = None)
+
+  case class Recipient(nodeId: PublicKey, secret: Option[ByteVector], padding: Option[ByteVector] = None)
+
+  def buildRoute(blindingSecret: PrivateKey,
+                 intermediateNodes: Seq[IntermediateNode],
+                 destination: Either[Recipient, Sphinx.RouteBlinding.BlindedRoute]): Sphinx.RouteBlinding.BlindedRoute = {
+    val last = destination match {
+      case Left(Recipient(nodeId, _, _)) => EncryptedRecipientDataTlv.OutgoingNodeId(nodeId) :: Nil
+      case Right(Sphinx.RouteBlinding.BlindedRoute(nodeId, blindingKey, _)) => EncryptedRecipientDataTlv.OutgoingNodeId(nodeId) :: EncryptedRecipientDataTlv.NextBlinding(blindingKey) :: Nil
+    }
+    val intermediatePayloads =
+      if (intermediateNodes.isEmpty) {
+        Nil
+      } else {
+        (intermediateNodes.tail.map(node => EncryptedRecipientDataTlv.OutgoingNodeId(node.nodeId) :: Nil) :+ last)
```

Comment:
I use `tail` because I want the next node ids. I'm zipping `intermediateNodes.tail` and `intermediateNodes` to get a tuple (next node, current node).
`intermediateNodes` contains all the nodes that are between us and the destination. It can be empty if we contact the destination directly.

Reply:
Right, that makes sense.